File: 887_eggDrop.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    int superEggDrop(int K, int N) {
        vector<vector<int>> dp(K+1, vector<int>(N+1, 0));
        return helper(dp, K, N);
    }

    int helper(vector<vector<int>>& dp, int K, int N) {
        if (1 == K) return N;
        if (0 == N) return 0;
        if (dp[K][N]!=0) return dp[K][N];

        int res = INT_MAX;
        int l=1, r=N+1; // broken 单调递增（i-1: 随i增，单调递增）；notBroken 单调递减（N-i: 随i增，单调递减）。因此这里可以用二分来求相交点
        while (l < r) {
            int mid = (l+r)/2;
            int broken = helper(dp, K-1, mid-1);
            int notBroken = helper(dp, K, N-mid);

            if (broken > notBroken) {
                r=mid;
                res = min(res, broken+1);
            } else {
                l=mid+1;
                res = min(res, notBroken+1);
            }
        }

        dp[K][N] = res;
        return res;
    }
}

File: dp/887_eggDrop.py
class Solution:
    def __init__(self):
        self.memo = dict()

    def superEggDrop(self, K: int, N: int):
        return self.dp(K, N)

    def dp(self, K, N) -> int:
        # base case
        if K == 1:
            return N
        if N == 0:
            return 0
        # 避免重复计算
        if (K, N) in self.memo:
            return self.memo[(K, N)]

        res = float('INF')
        # 穷举所有可能的选择
        for i in range(1, N + 1):
            res = min(res, max(self.dp(K, N - i), self.dp(K - 1, i - 1)) + 1)

        # 记入备忘录
        self.memo[(K, N)] = res
        return res


if __name__ == "__main__":
    obj = Solution()
    res = obj.superEggDrop(2, 10)
    print(res)
